src/aiortc/jitterbuffer.py

```diff
@@ -18,35 +19,44 @@ def __init__(self, capacity: int, prefetch: int = 0) -> None:
         self._origin: Optional[int] = None
         self._packets: List[Optional[RtpPacket]] = [None for i in range(capacity)]
         self._prefetch = prefetch
+        self.__max_number = 65536
 
     @property
     def capacity(self) -> int:
         return self._capacity
 
-    def add(self, packet: RtpPacket) -> Optional[JitterFrame]:
+    def add(self, packet: RtpPacket) -> Tuple[bool, Optional[JitterFrame]]:
+        pli_flag = False
         if self._origin is None:
             self._origin = packet.sequence_number
-        elif packet.sequence_number <= self._origin - MAX_MISORDER:
-            self.remove(self.capacity)
-            self._origin = packet.sequence_number
-        elif packet.sequence_number < self._origin:
-            return None
-
-        delta = packet.sequence_number - self._origin
-        if delta >= 2 * self.capacity:
-            # received packet is so far beyond capacity we cannot keep any
-            # previous packets, so reset the buffer
-            self.remove(self.capacity)
-            self._origin = packet.sequence_number
-        elif delta >= self.capacity:
-            # remove just enough packets to fit the received packets
+            delta = 0
+            misorder = 0
+        else:
+            delta = uint16_add(packet.sequence_number, -self._origin)
+            misorder = uint16_add(self._origin, -packet.sequence_number)
+
+        if misorder < delta:
+            if misorder >= MAX_MISORDER:
+                self.remove(self.capacity)
+                self._origin = packet.sequence_number
+                delta = misorder = 0
+                if self._capacity >= 128:
+                    pli_flag = True
+            else:
+                return pli_flag, None
+
+        if delta >= self.capacity:
+            # remove just enough frames to fit the received packets
             excess = delta - self.capacity + 1
-            self.remove(excess)
+            if self.smart_remove(excess):
```

Comment:
`smart_remove` seems to return a value.. which we ignore?

Reply:
We change it afterwards only when the buffer has been entirely purged (in such a case the smart_remove returns True). If we would delete the "self._origin = packet.sequence_number" from line 52, in case of situation where the buffer was purged we would and up with a self._origin value that points to a self._packets element that is None and this is a bad thing that would cause the buffer to be 100% filled no matter what (the remove_frame method would terminate on line 72 until the value would be overwritten).

Reply:
BTW we have the same situation with `remove` in line 40 and 41 of jitterbuffer.py. I've found that the previous code did not address the situation correctly when I've encountered some abnormal local network conditions, that caused a lot of packet drops.